program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,353 @@ use crate::c_oracle_header::{
+        let prev_entry = time_to_entry(prev_time, NUM_ENTRIES, self.granularity)?;
+        if current_entry != prev_entry && self.entry_validity[prev_entry] == Status::Pending {
+            self.entry_validity[prev_entry] = Status::Valid;
+        }
+        Ok(())
+    }
+    fn invalidate_intermediate_entries(
+        &mut self,
+        prev_time: i64,
+        current_time: i64,
+    ) -> Result<(), OracleError> {
+        let current_entry = time_to_entry(current_time, NUM_ENTRIES, self.granularity)?;
+        let prev_entry = time_to_entry(prev_time, NUM_ENTRIES, self.granularity)?;
+
+
+        let num_skipped_entries = current_entry - prev_entry;
+        let mut entry = prev_entry;
+        for _ in 0..min(num_skipped_entries, NUM_ENTRIES) {
+            entry = get_next_entry(entry, NUM_ENTRIES);
+            self.entry_validity[current_entry] = Status::Invalid;
+        }
+        Ok(())
+    }
+    fn update_current_tick(
+        &mut self,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        let current_entry = time_to_entry(current_time, NUM_ENTRIES, self.granularity)?;
+
+        self.prices[current_entry] = current_price;
+        self.confidences[current_entry] = current_conf;
+        self.entry_validity[current_entry] =
+            if get_time_to_entry_end(current_time, self.granularity) >= self.threshold {
+                Status::Invalid
+            } else {
+                Status::Pending
+            };
+        Ok(())
+    }
+}
+
+const THIRTY_MINUTES: i64 = 30 * 60;
+const TEN_HOURS: usize = 10 * 60 * 60;
+const TWENTY_SECONDS: u64 = 20;
 #[derive(Debug, Clone, Copy)]
 #[repr(C)]
 /// this wraps multiple SMA and tick trackers, and includes all the state
 /// used by the time machine
 pub struct TimeMachineWrapper {
-    //Place holder with the size of the fields I am planning to add
-    place_holder: [u8; 1864],
+    sma_tracker:  SmaTracker<{ TEN_HOURS / (THIRTY_MINUTES as usize) }>,
+    tick_tracker: TickTracker<{ TEN_HOURS / (THIRTY_MINUTES as usize) }>,
+}
+
+impl TimeMachineWrapper {
+    pub fn initialize(&mut self) -> Result<(), OracleError> {
+        self.sma_tracker
+            .initialize(TWENTY_SECONDS, THIRTY_MINUTES)?;
+        self.tick_tracker
+            .initialize(try_convert(TWENTY_SECONDS)?, THIRTY_MINUTES)?;
+        Ok(())
+    }
+
+    pub fn add_price(
+        &mut self,
+        prev_time: i64,
+        current_time: i64,
+        last_two_prices: (i64, i64),
+        last_two_confs: (u64, u64),
+        slot_gap: u64,
+    ) -> Result<(), OracleError> {
+        self.sma_tracker.add_price(
+            prev_time,
+            current_time,
+            last_two_prices,
+            last_two_confs,
+            slot_gap,
+        )?;
+        self.tick_tracker.add_price(
+            prev_time,
+            current_time,
+            last_two_prices.0,
+            last_two_confs.0,
```

Comment:
wait these tuples are ordered as `(cur_price, prev_price)`? that seems like the opposite of what someone would expect

imo if you're trying to reduce the number of arguments to these functions, a better approach would be a struct with fields like `{price, time, conf}`

Reply:
Actually, they aren't ordered that way, and this is a bug.

Reply:
Right, sorry about that. It originally was stroing the prev_tick then I swithced to this